.github/workflows/interop-test.yml

```diff
@@ -1,32 +1,40 @@
 name: Interoperability Testing
-
 on:
   pull_request:
   push:
     branches:
-      - master
-
-concurrency:
-  group: ${{ github.workflow }}-${{ github.ref }}
-  cancel-in-progress: true
+      - "master"
 
 jobs:
-  # NOTE: during a pull request run, github creates a merge commit referenced in `github.sha`
-  # that merge commit is not a regular commit. You won't find it with a regular `git checkout SHA` and
-  # tools like `go get repo@SHA` won't find it.
-  #
-  # As a workaround, we generate a path to the actual pull request's commit, it looks like:
-  # `github.com/external-org/go-libp2p@latest-commit-on-their-branch`
-  run-ping-interop-cross-version:
-    uses: "libp2p/test-plans/.github/workflows/run-composition.yml@master"
-    with:
-      composition_file: "ping/_compositions/rust-cross-versions.toml"
-      custom_git_target: github.com/${{ github.event.pull_request.head.repo.full_name || github.event.repository.full_name }}
-      custom_git_reference: ${{ github.event.pull_request.head.sha || github.sha }}
-  run-ping-interop-cross-implementation:
-    uses: "libp2p/test-plans/.github/workflows/run-composition.yml@master"
+  build-ping-container:
+    name: Build Ping interop container
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          submodules: recursive
+      - name: Build image
+        working-directory: ./test-plans
+        run: make
+      - name: Upload ping versions info
+        uses: actions/upload-artifact@v3
+        with:
+          name: ping-versions
+          path: ./test-plans/ping-versions.json
+      - name: Upload image tar
+        uses: actions/upload-artifact@v3
+        with:
+          name: ping-image
+          path: ./test-plans/ping-image.tar
```

Comment:
Now that we are using plain docker compose, can't we just build the image as part of the workflow here, upload it to GitHub's container registry and reference it here? That seems a lot simpler than storing it as an artifact.

cc @MarcoPolo

Reply:
In more detail, why don't we have the following:

1. A workflow per repository that only runs on pushes to master that builds a docker container and uploads it to the container registry under the tag "master".
2. A workflow per repository that only runs on tag pushes and builds a docker container and uploads it to the container registry under the Git tag, i.e. `0.45.0`
3. A PR workflow that builds a container of the current commit, uploads it to the container registry under the tag of the PR number.

Once (3) finishes, we can then kick off a dependent workflow that runs a matrix of containers against each other.

Reply:
You can upload to a container registry if you want. If the containerImageID is a tag, then docker will automatically try to fetch that tag. You don't have to upload image artifacts if you don't want to.

That said, I think the approach of building the container from the commit is better. For one it's guaranteed to be reproducible. You don't have multiple commits from the same PR mutating the tag number. Two, it's the same workflow on master, PRs, or tags. Three, when you introduce a container registry, now you need to deal with authentication, limits, and garbage collecting from that container registry.

The one benefit I see to container registries is them acting as a cache for faster tests. But we can cache layers already using GitHub Actions cache. 

I'm not sure the extra work involved in setting up and maintaining a registry is worth it. But of course, if you want to feel free. I'm just sharing my thoughts from thinking about this as well.

Reply:
Thanks for sharing your thoughts. Perhaps a mix of both would be the best solution?

- The `master` tag would only ever produce one container so not much GC needed.
- The containers for releases are useful for other implementations too. I don't think we can utilize the GH Actions Cache here?
- Not uploading the container of the current commit/PR is a good idea, those would just be garbage we need to collect.

If we can remove container builds of master and tags from this workflow, then the Dockerfile also gets simpler because we always just build for the current source, no overrides needed.

Reply:
> I'm not sure the extra work involved in setting up and maintaining a registry is worth it. But of course, if you want to feel free. I'm just sharing my thoughts from thinking about this as well.

GitHub makes authenticating and interacting with its own registry really easy so that wouldn't be much work in my experience.

Reply:
I think the ability to checkout an arbitrary commit of test-plans and be able to run the exact those exact tests is a very good feature (reproduicibility). As long as we hold that, I don’t mind using tagged images in a registry.

But what is the benefit of tagging images in a registry? To me, the only benefit is faster builds. But you can get this with caching as well. Maybe standard GHA cache isn’t good enough (since it caches per branch I believe), but we could cache the layers from the latest test-plans master builds as well, and I think that would cover all the use cases of tagged images.

> The master tag would only ever produce one container so no GC needed, i.e. the container for master would be updated every time the master branch gets updated.

I don’t see the benefit here since you are building this every time. When do you get to reuse this tag?

> If we can remove container builds of master and tags from this workflow, then the Dockerfile also gets simpler because we always just build for the current source, no overrides needed.

What overrides are used? I imagine you are just building from the current source.

One idea for released versions is that inside the test-plans repo we specify the way to build the interop image is to checkout this repo at this commit, then cd into test-plans and run `make`.

Reply:
> I think the ability to checkout an arbitrary commit of test-plans and be able to run the exact those exact tests is a very good feature (reproduicibility). As long as we hold that, I don’t mind using tagged images in a registry.
> 
> But what is the benefit of tagging images in a registry? To me, the only benefit is faster builds.

The benefit is faster builds but it also just feels wrong. The source code for a test for e.g. libp2p 0.45 should _never_ change. The version is out there, our users are using it. If we want to take interoperability seriously, I think it is good to constrain ourselves such that changing the test for a particular released version is _hard_.

Building the binary every time doesn't sit well for me on that front.

> Maybe standard GHA cache isn’t good enough (since it caches per branch I believe), but we could cache the layers from the latest test-plans master builds as well, and I think that would cover all the use cases of tagged images.

It caches per user-defined key which can be completely arbitrary and is not restricted to branches or anything. Bear in mind that each repo only has 10GB here and at least in `rust-libp2p` we already have to jump through hoops to not blow this cache limit for our regular jobs :(

> > The master tag would only ever produce one container so no GC needed, i.e. the container for master would be updated every time the master branch gets updated.
> 
> I don’t see the benefit here since you are building this every time. When do you get to reuse this tag?

It gets reused across all pushes on all open PRs until the next one merges. I myself have usually anywhere from 2-10 PRs open that I work on in parallel i.e. I switch to another PR once I incorporated feedback and wait for the next round of reviews.

> > If we can remove container builds of master and tags from this workflow, then the Dockerfile also gets simpler because we always just build for the current source, no overrides needed.
> 
> What overrides are used? I imagine you are just building from the current source.

My bad, I wrongly assumed that we would still need the "sed"-ing inside the Dockerfile. That should be gone now in any case.

> One idea for released versions is that inside the test-plans repo we specify the way to build the interop image is to checkout this repo at this commit, then cd into test-plans and run `make`.

This exact step could have happened when the tag was pushed. It is like pre-built binaries of any other application here on GitHub.

Reply:
Thanks. I think we're starting to understand each other's points and we agree on the fundamentals.

Just a couple of questions for me:

> It gets reused across all pushes on all open PRs until the next one merges.

Maybe there's a misunderstanding somewhere? An open rust-libp2p PR wouldn't do an interop test against rust-libp2p master (we could do that, I'm just not sure it's useful). It runs interop tests against every other released version of libp2p.

>  The source code for a test for e.g. libp2p 0.45 should never change. The version is out there, our users are using it. If we want to take interoperability seriously, I think it is good to constrain ourselves such that changing the test for a particular released version is hard.

I agree with this. I don't think building from source prevents this. If we constrain ourselves to never mutate an image tag, then I think we agree except on the little details. To me a sha256 hash is by definition an immutable tag of an image.

I don't have strong opinions on how rust-libp2p wants to handle building their image. If the project wants to simply submit a container tag that will be used in the interop tests, I only ask that the tag is immutable.

Reply:
> > It gets reused across all pushes on all open PRs until the next one merges.
> 
> Maybe there's a misunderstanding somewhere? An open rust-libp2p PR wouldn't do an interop test against rust-libp2p master (we could do that, I'm just not sure it's useful). It runs interop tests against every other released version of libp2p.

Interesting, I thought we also ran interop tests against our own master branch. @mxinden do I remember this incorrectly?

If not then I agree that it is not useful to a moving `master` tag.

> >  The source code for a test for e.g. libp2p 0.45 should never change. The version is out there, our users are using it. If we want to take interoperability seriously, I think it is good to constrain ourselves such that changing the test for a particular released version is hard.
> 
> I agree with this. I don't think building from source prevents this. If we constrain ourselves to never mutate an image tag, then I think we agree except on the little details. To me a sha256 hash is by definition an immutable tag of an image.
> 
> I don't have strong opinions on how rust-libp2p wants to handle building their image. If the project wants to simply submit a container tag that will be used in the interop tests, I only ask that the tag is immutable. 

The tag should definitely be immutable. As far as I know, one can append the sha256 hash of a docker image tag to the overall ID and thus guarantee that anyone pulling this image from the registry will get exactly the image with this hash. That is how I would specify those images in test-plans.

Example:

`docker.io/library/hello-world@sha256:faa03e786c97f07ef34423fccceeec2398ec8a5759259f94d99078f264e9d7a`

Reply:
> > > It gets reused across all pushes on all open PRs until the next one merges.
> > 
> > 
> > Maybe there's a misunderstanding somewhere? An open rust-libp2p PR wouldn't do an interop test against rust-libp2p master (we could do that, I'm just not sure it's useful). It runs interop tests against every other released version of libp2p.
> 
> Interesting, I thought we also ran interop tests against our own master branch. @mxinden do I remember this incorrectly?

If I recall correctly, we did. I was planning to have each pull request tested against `master` with the wo-testground tests, though based on intuition, without specific reasoning. Thinking about it some more, I don't see what kind of bugs this would catch, thus I suggest not investing into PR-to-master testing for now. Thoughts?